### PR TITLE
drivers: dma_lpc: Suppport dest_scatter and source_gather

### PR DESCRIPTION
**Pull request by mmahadevan108**
Process dest_scatter_interval and source_gather_interval configurations and accordingly set the source and destination increment values.

### CODE REVIEW

File: drivers/dma/dma_mcux_lpc.c [hunk 1]
```diff
@@ -41,6 +41,8 @@ struct channel_data {
 	void *user_data;
 	dma_callback_t dma_callback;
 	enum dma_channel_direction dir;
+	uint8_t src_inc;
```
Comment (danieldegrasse) [inline drivers/dma/dma_mcux_lpc.c:44]: We should clear these values in the `dma_mcux_lpc_clear_channel_data` function

Comment (mmahadevan108) [reply drivers/dma/dma_mcux_lpc.c:44]: Added this code.

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/dma/dma_mcux_lpc.c [hunk 2]
```diff
@@ -340,16 +353,60 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
...
 	case MEMORY_TO_MEMORY:
 		is_periph = false;
-		src_inc = 1;
-		dst_inc = 1;
+		if (block_config->source_gather_en) {
+			src_inc = block_config->source_gather_interval / width;
+			/* The current controller only supports incrementing the
+			 * source and destination up to 4 time transfer width
+			 */
+			if ((src_inc > 4) || (src_inc == 3)) {
```
Comment (decsny) [inline drivers/dma/dma_mcux_lpc.c:361]: https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#ga52d277cbf33f76350b2fcb21c24640ee
`IS_POWER_OF_TWO` I am assuming that's what the check for 3 is

File: drivers/dma/dma_mcux_lpc.c [hunk 3]
```diff
@@ -281,7 +285,7 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 	struct dma_block_config *block_config;
 	uint32_t virtual_channel;
 	uint8_t otrig_index;
-	uint8_t src_inc, dst_inc;
+	uint8_t src_inc = 0, dst_inc = 0;
```
Comment (decsny) [inline drivers/dma/dma_mcux_lpc.c:288]: maybe these can start as 1. then the else blocks can be removed below

File: drivers/dma/dma_mcux_lpc.c [hunk 4]
```diff
@@ -642,26 +702,11 @@ static int dma_mcux_lpc_reload(const struct device *dev, uint32_t channel,
...
-	case MEMORY_TO_MEMORY:
-		src_inc = 1;
-		dst_inc = 1;
-		break;
-	case MEMORY_TO_PERIPHERAL:
-		src_inc = 1;
-		dst_inc = 0;
-		break;
-	case PERIPHERAL_TO_MEMORY:
-		src_inc = 0;
```
Comment (decsny) [inline drivers/dma/dma_mcux_lpc.c:658]: I am a little confused how we are handling peripheral to memory and vice versa now?

File: drivers/dma/dma_mcux_lpc.c (hunk 2 above)
Comment (mmahadevan108) [reply drivers/dma/dma_mcux_lpc.c:361]: The only values we allow are 1, 2, 4. The macro you provided will work as well. However checking for 3 in this case seems simpler.

File: drivers/dma/dma_mcux_lpc.c (hunk 3 above)
Comment (mmahadevan108) [reply drivers/dma/dma_mcux_lpc.c:288]: Good idea, I will make this change.

File: drivers/dma/dma_mcux_lpc.c (hunk 4 above)
Comment (mmahadevan108) [reply drivers/dma/dma_mcux_lpc.c:658]: We would always increment by `1 x width`.

**Review by danieldegrasse — DISMISSED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Review by decsny — APPROVED** [review]

File: drivers/dma/dma_mcux_lpc.c [hunk 5]
```diff
@@ -706,8 +744,8 @@ static int dma_mcux_lpc_reload(const struct device *dev, uint32_t channel,
 		/* Only one buffer, enable interrupt */
 		xfer_config = DMA_CHANNEL_XFER(0UL, 0UL, 1UL, 0UL,
 					data->width,
-					src_inc,
-					dst_inc,
+					data->src_inc,
```
Comment (decsny) [inline drivers/dma/dma_mcux_lpc.c:747]: Is it the same for every channel?

**Pull request merged**